geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/builder/AbstractFeatureTypeBuilder.scala

```diff
@@ -0,0 +1,56 @@
+package org.locationtech.geomesa.utils.geotools.builder
+
+import org.geotools.feature
+import org.geotools.feature.NameImpl
+import org.geotools.feature.`type`.{ComplexTypeImpl, FeatureTypeImpl}
+import org.geotools.gml3.GMLSchema
+import org.opengis.feature.`type`.{ComplexType, Name, PropertyDescriptor, PropertyType, AttributeDescriptor}
+
+import scala.collection.Iterator.iterate
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+object AbstractFeatureTypeBuilder {
+  def isDescendedFrom(child: PropertyType, ancestor: PropertyType): Boolean = iterate(child)(_.getSuper).takeWhile(_ != null).contains(ancestor)
+  def subclassByRestriction(child: PropertyType) =
+    isDescendedFrom(child, GMLSchema.FEATUREPROPERTYTYPE_TYPE)
+}
+
+abstract class AbstractFeatureTypeBuilder[ReturnType] {
+
+
+  import AbstractFeatureTypeBuilder._
+  var targetNs: String = null
+  var _name: NameImpl = null
+  var superType: ComplexType = null
+  var descIndex = 0
+  final val descriptors: mutable.Buffer[PropertyDescriptor] = mutable.Buffer.empty
+
+  def namespace(ns: String): this.type = {
+    targetNs = ns
+    this
+  }
+
+  def `type`(name: String): DescriptorBuilder[_<:ReturnType] = `type`(name, GMLSchema.ABSTRACTFEATURETYPE_TYPE)
+
+  def `type`(name: String, superType: ComplexType): DescriptorBuilder[ReturnType] = {

```

Comment:
Hmm...  are there any suggestions/possibilities for getting around naming the function `type`.  

'typename' seems to fit for the first use, but it seems sloppy for the second.